### DISCUSSION TITLE
Tank Demo Aiming

### DISCUSSION THREAD

**Discussion by 7Bpencil** [bepu/bepuphysics2#117] [Q&A] [answered]
Hello Norbo! I fiddled around with the tank demo, and now I am trying to point a tank's barrel on a target. The problem is it's slightly inaccurate unless the angle between body and barrel equals 0 or 90 degrees:

![bandicam 2020-12-30 18-31-51-064](https://user-images.githubusercontent.com/43344410/103355694-b24d5e80-4ad0-11eb-8222-f6191d29d811.png)

For a test, I placed the sphere of radius 1 in position (3, 3, 3)
Aiming code:
```cs
var targetPosition = new Vector3(3);
var turretBody = Simulation.Bodies.GetBodyReference(playerController.Tank.Turret);
var aimDirection = Vector3.Normalize(targetPosition - turretBody.Pose.Position);
playerController.UpdateMovementAndAim(Simulation, leftTargetSpeedFraction, rightTargetSpeedFraction, 
				      zoom, brake, brake, aimDirection);
```

My first idea was that offset between turret origin and turret anchor causes the problem, but it's not the case. (because this version produces same results:)
```cs
var targetPosition = new Vector3(3);
var turretBody = Simulation.Bodies.GetBodyReference(playerController.Tank.Turret);
RigidPose.Transform(in playerController.Tank.TurretLocalTurretAnchor, 
                    turretBody.Pose, 
                    out var turretAnchorWorldSpace);
var aimDirection = Vector3.Normalize(targetPosition - turretAnchorWorldSpace);
playerController.UpdateMovementAndAim(Simulation, leftTargetSpeedFraction, rightTargetSpeedFraction, 
		                      zoom, brake, brake, aimDirection);
```

What do you think?

**Comment by RossNordby** [comment, marked as answer]
MathHelper.ApproximateAcos, the vectorized acos approximation used by the involved constraints, was a little too approximate. Try out 4b72cd62ec9592294e68389362ef747815ad04ba; it should help.

It's still using a very Ye Olde implementation and could be a wee bit better, but the maximum error is about 1000x lower now. I might revamp it again once I can use intrinsics.

**Comment by 7Bpencil** [comment]
Thank you! Now it's juuuuust perfect. Happy New Year!
![wewfe](https://user-images.githubusercontent.com/43344410/103423638-8235b680-4bc9-11eb-8e67-e17c1d3ed51a.png)
(Also I would like to see beta11 in NuGet)

**Comment by RossNordby** [comment]
Should be up- I did forget that nuget orders 2.3.0-beta11 as *before* 2.3.0-beta9, though, so you have to manually select it. Oopsy.

**Reply by 7Bpencil** [reply]
haha lexicographic order can confuse even NuGet
![Сниwcwecмок](https://user-images.githubusercontent.com/43344410/103425458-3c322000-4bd4-11eb-8f6d-badb7f849e6f.PNG)
